source/src/ble/cs_CrownstoneCentral.cpp

```diff
@@ -0,0 +1,663 @@
+		return;
+	}
+
+	// In theory, we can receive notifications at any moment.
+	// So only handle them when we expect them, and ignore them otherwise.
+	if (_currentOperation == Operation::WRITE && static_cast<WriteControlSteps>(_currentStep) == WriteControlSteps::RECEIVE_RESULT) {
+
+		cs_data_t resultData;
+		cs_ret_code_t retCode = mergeNotification(result.data, resultData);
+		switch (retCode) {
+			case ERR_SUCCESS: {
+				if (resultData.data == nullptr) {
+					finalizeOperation(_currentOperation, ERR_BUFFER_UNASSIGNED);
+					return;
+				}
+				// Check the result data.
+				ResultPacketAccessor<> resultPacketAccessor;
+				retCode = resultPacketAccessor.assign(resultData.data, resultData.len);
+				if (retCode != ERR_SUCCESS) {
+					finalizeOperation(_currentOperation, retCode);
+					return;
+				}
+				_log(LogLevelCsCentralDebug, false, "Result: protocol=%u type=%u result=%u data=",
+						resultPacketAccessor.getProtocolVersion(),
+						resultPacketAccessor.getType(),
+						resultPacketAccessor.getResult());
+				_logArray(LogLevelCsCentralDebug, true, resultPacketAccessor.getPayload().data, resultPacketAccessor.getPayload().len);
+
+				cs_central_write_result_t result = {
+						.writeRetCode = ERR_SUCCESS,
+						.result = resultPacketAccessor
+				};
+				if (resultPacketAccessor.getResult() == ERR_WAIT_FOR_SUCCESS) {
+					// Send event, but keep waiting for the next result.
+					// No need to check the current operation or step, that has been earlier in this function.
+					event_t resultEvent(CS_TYPE::EVT_CS_CENTRAL_WRITE_RESULT, reinterpret_cast<uint8_t*>(&result), sizeof(result));
+					resultEvent.dispatch();
+					return;
+				}
+				// Done.
+				finalizeOperation(_currentOperation, reinterpret_cast<uint8_t*>(&result), sizeof(result));
+				break;
+			}
+			case ERR_WAIT_FOR_SUCCESS: {
+				// Keep waiting for more notifications.
+				break;
+			}
+			default: {
+				finalizeOperation(_currentOperation, retCode);
+				break;
+			}
+		}
+	}
+}
+
+cs_ret_code_t CrownstoneCentral::mergeNotification(const cs_const_data_t& data, cs_data_t& resultData) {
+	// Check data size.
+	const uint16_t headerSize = 1;
+	if (data.len < (headerSize + 1)) {
+		LOGw("Invalid length=%u", data.len);
+		return ERR_WRONG_PAYLOAD_LENGTH;
+	}
+	const uint16_t dataSize = data.len - headerSize;
+
+	// Check index.
+	uint8_t index = data.data[0];
+	if (index != _notificationNextIndex && index != 255) {
+		LOGw("Expected index=%u, received index=%u", _notificationNextIndex, index);
+		return ERR_WRONG_PARAMETER;
+	}
+
+	// Get the encryption buffer.
+	cs_data_t encryptedBuffer;
+	if (!EncryptionBuffer::getInstance().getBuffer(encryptedBuffer.data, encryptedBuffer.len)) {
+		return ERR_BUFFER_UNASSIGNED;
+	}
+
+	if (encryptedBuffer.len < _notificationMergedDataSize + dataSize) {
+		return ERR_BUFFER_TOO_SMALL;
+	}
+
+	// Copy data to encryption buffer.
+	memcpy(encryptedBuffer.data + _notificationMergedDataSize, data.data + headerSize, dataSize);
```

Comment:
I saw memmove somewhere else. Are we going to memmove where possible or is memcpy still an acceptable option?

Reply:
I started using memmove when the chances are that the source buffer overlaps with the destination buffer.
In this case, the data comes from some SD memory, it will never overlap with the encryption buffer. If it would, storing the merged data in the encryption buffer wouldn't work at all.